src/Optimizer/Transformer/AutoExtensions.php

```diff
@@ -0,0 +1,584 @@
+             * bindvalue).
+             */
+            if (
+                $this->configuration->get(AutoExtensionsConfiguration::EXPERIMENT_BIND_ATTRIBUTE)
+                && strpos($attribute->name, self::BIND_SHORT_FORM_PREFIX) === 0
+            ) {
+                // Change name from bindaria-labelledby to aria-labelledby.
+                $attributeNameWithoutBindPrefix = substr($attribute->name, strlen(self::BIND_SHORT_FORM_PREFIX));
+
+                // Change the name from aria-labelledby to [ARIA_LABELLEDBY].
+                $bindableAttributeName = '['
+                    . preg_replace('/\-/', '_', strtoupper($attributeNameWithoutBindPrefix))
+                    . ']';
+
+                // Rename attribute from bindx to data-amp-bind-x.
+                if ($globalAttributes->has($bindableAttributeName)) {
+                    $newAttributeName = Amp::BIND_DATA_ATTR_PREFIX . $attributeNameWithoutBindPrefix;
+                    $newAttributes[$newAttributeName] = $attribute->value;
+
+                    $node->removeAttribute($attribute->name);
+
+                    $usesAmpBind = true;
+                }
+            }
+        }
+
+        // Add the renamed bindable attributes in the node.
+        if ($newAttributes) {
+            $node->setAttributes($newAttributes);
+        }
+
+        if ($usesAmpBind) {
+            $extensionScripts = $this->maybeAddExtension($document, $extensionScripts, Extension::BIND);
+        }
+
+        return $extensionScripts;
+    }
+
+    /**
+     * Get all attributes for a tag.
+     *
+     * @param string $tagName Name of the node tag.
+     * @return array Attribute list for the tag name.
+     */
+    private function getTagAttributeList($tagName)
+    {
+        static $nodeAttributeList = [];
+
+        if (! isset($nodeAttributeList[$tagName])) {
+            $nodeAttributeList[$tagName] = [];
+
+            $tagSpecs = $this->spec->tags()->byTagName($tagName);
+
+            foreach ($tagSpecs as $tagSpec) {
+                $attrLists = $tagSpec->get(SpecRule::ATTR_LISTS);
+                if (is_array($attrLists)) {
+                    foreach ($attrLists as $attrList) {
+                        $list = $this->spec->attributeLists()->get($attrList);
+                        $nodeAttributeList[$tagName] = array_merge($nodeAttributeList[$tagName], $list::ATTRIBUTES);
+                    }
+                }
+            }
+        }
+
+        return $nodeAttributeList[$tagName];
+    }
+
+    /**
+     * Remove unneeded extension scripts.
+     *
+     * @param Document  $document         Document to scan for unneeded extensions.
+     * @param Element[] $extensionScripts Array of extension scripts to check.
+     * @return Element[] Adapted array of extension scripts.
+     */
+    private function removeUnneededExtensions(Document $document, $extensionScripts)
+    {
+        // TODO: Add logic.
```

Comment:
This part is still missing the implementation. It should probably also be configurable via a flag, so you can choose whether removal happens or not.

Reply:
Aren't ignoredExtensions and removeExtensions the same? Cause we are checking the ignoredExtensions list before adding that extension into `extensionScripts` list. For example if we want to remove `amp-form` extension, we can put this into ignoredExtensions list. Please share your thoughts on this.

Reply:
Ignoring an extension means that the system should not check what to do with that extension. If the extension script is present, it stays in (whether it is needed or not). If the extension script is not present, it is not being added (whether it is needed or not).

Removing unneeded extensions on the other hand is about doing cleanup in case their are left-overs. For an existing document that might already have extension scripts, you can have both missing extension scripts (the author forgot one) and superfluous extension scripts (the author added one, but later removed the actual markup but forgot to remove the corresponding extension script as well). So we are already adding the extension scripts that are missing, but we are not yet removing the extension scripts that are unused.

Ignore => do not even do a check
Remove => strip if unused